From feb5267852bb7e4cd304255c7ad36f590e796aa2 Mon Sep 17 00:00:00 2001
From: Alexandre Julliard <julliard@winehq.org>
Date: Wed, 4 Feb 2026 14:31:59 +0100
Subject: [PATCH] winecrt0: Add a helper for loading a unixlib by name.

(cherry picked from commit 39a9fa263d3ea016105f04cb034e66223f1708d1)

CW-Bug-Id: #27687
---
 dlls/winecrt0/unix_lib.c | 19 +++++++++++++++++++
 include/wine/unixlib.h   |  3 +++
 2 files changed, 22 insertions(+)

diff --git a/dlls/winecrt0/unix_lib.c b/dlls/winecrt0/unix_lib.c
index e2eb079bd48..ce0cb3b0c22 100644
--- a/dlls/winecrt0/unix_lib.c
+++ b/dlls/winecrt0/unix_lib.c
@@ -94,3 +94,22 @@ NTSTATUS WINAPI __wine_init_unix_call(void)
     return NtQueryVirtualMemory( GetCurrentProcess(), image_base(), MemoryWineLoadUnixLib,
                                  &__wine_unixlib_handle, sizeof(__wine_unixlib_handle), NULL );
 }
+
+NTSTATUS WINAPI __wine_load_unix_lib( const UNICODE_STRING *name, unixlib_module_t *lib,
+                                      unixlib_handle_t *handle )
+{
+    UINT64 res[2];
+    NTSTATUS status = NtQueryVirtualMemory( GetCurrentProcess(), name, MemoryWineLoadUnixLibByName,
+                                            res, handle ? sizeof(res) : sizeof(res[0]), NULL );
+    if (!status)
+    {
+        if (lib) *lib = res[0];
+        if (handle) *handle = res[1];
+    }
+    return status;
+}
+
+NTSTATUS WINAPI __wine_unload_unix_lib( unixlib_module_t lib )
+{
+    return NtQueryVirtualMemory( GetCurrentProcess(), &lib, MemoryWineUnloadUnixLib, NULL, 0, NULL );
+}
diff --git a/include/wine/unixlib.h b/include/wine/unixlib.h
index ddd28b24e24..77f9c7cb374 100644
--- a/include/wine/unixlib.h
+++ b/include/wine/unixlib.h
@@ -275,6 +275,9 @@ NTSYSAPI int ntdll_wcsnicmp( const WCHAR *str1, const WCHAR *str2, int n );
 extern unixlib_handle_t __wine_unixlib_handle;
 extern NTSTATUS (WINAPI *__wine_unix_call_dispatcher)( unixlib_handle_t, unsigned int, void * );
 extern NTSTATUS WINAPI __wine_init_unix_call(void);
+extern NTSTATUS WINAPI __wine_load_unix_lib( const UNICODE_STRING *name, unixlib_module_t *lib,
+                                             unixlib_handle_t *handle );
+extern NTSTATUS WINAPI __wine_unload_unix_lib( unixlib_module_t lib );
 
 #ifdef __arm64ec__
 NTSTATUS __wine_unix_call_arm64ec( unixlib_handle_t handle, unsigned int code, void *args );
